File: libraries/core_network/src/main/java/com/tokopedia/core/discovery/model/Filter.java
package com.tokopedia.core.discovery.model;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.Gson;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.List;

/**
 * @author kulomady on 12/22/16.
 */
public class Filter implements Parcelable {

    public static final String TEMPLATE_NAME_LOCATION = "template_location";
    public static final String TEMPLATE_NAME_OTHER = "template_other";
    
    private static final String TEMPLATE_NAME_SEPARATOR = "template_separator";
    private static final String TEMPLATE_NAME_RATING = "template_rating";
    private static final String TEMPLATE_NAME_SIZE = "template_size";
    private static final String TEMPLATE_NAME_CATEGORY = "template_category";
    private static final String TEMPLATE_NAME_COLOR = "template_color";
    private static final String TEMPLATE_NAME_PRICE = "template_price";
    private static final String TEMPLATE_NAME_BRAND = "template_brand";

    @SerializedName("title")
    @Expose
    String title;
    @SerializedName("template_name")
    @Expose
    String templateName;
    @SerializedName("search")
    @Expose
    Search search;
    @SerializedName("options")
    @Expose
    List<Option> options = new ArrayList<>();

    public Filter() {
    }

    public boolean isSeparator() {
        return TEMPLATE_NAME_SEPARATOR.equals(templateName);
    }

    public boolean isCategoryFilter() {
        return TEMPLATE_NAME_CATEGORY.equals(templateName);
    }

    public boolean isColorFilter() {
        return TEMPLATE_NAME_COLOR.equals(templateName);
    }

    public boolean isPriceFilter() {
        return TEMPLATE_NAME_PRICE.equals(templateName);
    }

    public boolean isRatingFilter() {
        return TEMPLATE_NAME_RATING.equals(templateName);
    }

    public boolean isSizeFilter() {
        return TEMPLATE_NAME_SIZE.equals(templateName);
    }

    public boolean isLocationFilter() {
        return TEMPLATE_NAME_LOCATION.equals(templateName);
    }

    public boolean isBrandFilter() {
        return TEMPLATE_NAME_BRAND.equals(templateName);
    }

    public boolean isOtherFilter() {
        return TEMPLATE_NAME_OTHER.equals(templateName);
    }

    public boolean isExpandableFilter() {
        return isCategoryFilter() || isColorFilter() || isRatingFilter()
                || isSizeFilter() || isBrandFilter() || isLocationFilter()
                || isOtherFilter() || options.size() > 1;
    }

    /**
     * @return The title
     */
    public String getTitle() {
        return title;
    }

    /**
     * @param title The title
     */
    public void setTitle(String title) {
        this.title = title;
    }

    public String getTemplateName() {
        return templateName;
    }

    public void setTemplateName(String templateName) {
        this.templateName = templateName;
    }

    /**
     * @return The search
     */
    public Search getSearch() {
        return search;
    }

    /**
     * @param search The search
     */
    public void setSearch(Search search) {
        this.search = search;
    }

    /**
     * @return The options
     */
    public List<Option> getOptions() {
        if (options == null) {
            options = new ArrayList<>();
        }
        return options;
    }

    /**
     * @param options The options
     */
    public void setOptions(List<Option> options) {
        this.options = options;
    }

    @Override
    public String toString() {
        return new Gson().toJson(this);
    }


    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.title);
        dest.writeString(this.templateName);
        dest.writeParcelable(this.search, flags);
        dest.writeTypedList(this.options);
    }

    protected Filter(Parcel in) {
        this.title = in.readString();
        this.templateName = in.readString();
        this.search = in.readParcelable(Search.class.getClassLoader());
        this.options = in.createTypedArrayList(Option.CREATOR);
    }

    public static final Creator<Filter> CREATOR = new Creator<Filter>() {
        @Override
        public Filter createFromParcel(Parcel source) {
            return new Filter(source);
        }

        @Override
        public Filter[] newArray(int size) {
            return new Filter[size];
        }
    };
}

File: features/discovery/filter/src/main/java/com/tokopedia/filter/common/data/Filter.kt
package com.tokopedia.filter.common.data

import android.os.Parcelable
import com.google.gson.Gson
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import kotlinx.android.parcel.Parcelize
import java.util.ArrayList

@Parcelize
class Filter(@SerializedName("title")
             @Expose
             var title: String = "",

             @SerializedName("subTitle", alternate = ["subtitle"])
             @Expose
             var subTitle: String = "",

             @SerializedName("template_name", alternate = ["templateName"])
             @Expose
             var templateName: String = "",

             @SerializedName("search")
             @Expose
             var search: Search = Search(),

             @SerializedName("isNew")
             @Expose
             var isNew: Boolean = false,

             @SerializedName("filter_attribute_detail")
             @Expose
             var filterAttributeDetail: String = "",

             @SerializedName("chip_name")
             @Expose
             var chipName: String = "",

             @SerializedName("options")
             @Expose
             var options: List<Option> = ArrayList()) : Parcelable {

    fun clone(
            title: String? = null,
            subTitle: String? = null,
            templateName: String? = null,
            search: Search? = null,
            isNew: Boolean? = null,
            filterAttributeDetail: String? = null,
            options: List<Option>? = null,
    ): Filter {
        return Filter(
                title = title ?: this.title,
                subTitle = subTitle ?: this.subTitle,
                templateName = templateName ?: this.templateName,
                search = search ?: this.search,
                isNew = isNew ?: this.isNew,
                filterAttributeDetail = filterAttributeDetail ?: this.filterAttributeDetail,
                options = options ?: this.options,
        )
    }

    val isSeparator: Boolean
        get() = TEMPLATE_NAME_SEPARATOR.equals(templateName)

    val isCategoryFilter: Boolean
        get() = TEMPLATE_NAME_CATEGORY.equals(templateName)

    val isColorFilter: Boolean
        get() = TEMPLATE_NAME_COLOR.equals(templateName)

    val isOfferingFilter: Boolean
        get() = TEMPLATE_NAME_OFFERING.equals(templateName)

    val isPriceFilter: Boolean
        get() = TEMPLATE_NAME_PRICE.equals(templateName)

    val isRatingFilter: Boolean
        get() = TEMPLATE_NAME_RATING.equals(templateName)

    val isSizeFilter: Boolean
        get() = TEMPLATE_NAME_SIZE.equals(templateName)

    val isLocationFilter: Boolean
        get() = TEMPLATE_NAME_LOCATION.equals(templateName)

    val isBrandFilter: Boolean
        get() = TEMPLATE_NAME_BRAND.equals(templateName)

    val isOtherFilter: Boolean
        get() = TEMPLATE_NAME_OTHER.equals(templateName)

    val isExpandableFilter: Boolean
        get() = (isCategoryFilter || isColorFilter || isRatingFilter
                || isSizeFilter || isBrandFilter || isLocationFilter
                || isOtherFilter || options.size > 1)

    val isKeywordFilter: Boolean
        get() = TEMPLATE_NEGATIVE_KEYWORD == templateName

    val isPriceRangeCheckboxFilter: Boolean
        get() = TEMPLATE_PRICING_FOOD.equals(templateName)

    fun getFlattenedOptions() : List<Option> {
        return options.flatMap { option ->
            option.levelTwoCategoryList.flatMap { levelTwoCategory ->
                levelTwoCategory.levelThreeCategoryList.map { levelThreeCategory ->
                    levelThreeCategory.asOption()
                } + levelTwoCategory.asOption()
            } + option
        }
    }

    override fun toString(): String {
        return Gson().toJson(this)
    }

    companion object {

        const val TEMPLATE_NAME_LOCATION = "template_location"
        const val TEMPLATE_NAME_OTHER = "template_other"
        const val TEMPLATE_NAME_SEPARATOR = "template_separator"
        const val TEMPLATE_NAME_RATING = "template_rating"
        const val TEMPLATE_NAME_SIZE = "template_size"
        const val TEMPLATE_NAME_CATEGORY = "template_category"
        const val TEMPLATE_NAME_COLOR = "template_color"
        const val TEMPLATE_NAME_PRICE = "template_price"
        const val TEMPLATE_NAME_BRAND = "template_brand"
        const val TEMPLATE_NAME_OFFERING = "template_offer"
        const val TEMPLATE_NEGATIVE_KEYWORD = "template_negative_keyword"
        const val TEMPLATE_PRICING_FOOD = "template_pricing_food"
    }
}
